perf(ast/estree): speed up building UTF8-UTF16 translation table with SIMD

overlookmotel opened the pull request:
Use SIMD operations to speed up building the table of UTF8-UTF16 offset translations.

Previously, we looped through the source byte-by-byte, checking if each byte is ASCII, one-by-one. Instead, use SIMD to check blocks of 32 bytes at a time, and only drop down to byte-by-byte processing if the quick check does find any non-ASCII bytes in that block.

Building the table is about 7x faster in our benchmark after this PR.

Note: There are no explicit SIMD operations involved in the code. Just the code is written so that the compiler can auto-vectorize it.

We could get even better performance if we did use SIMD proper - on x86_64 with AVX-512 or Apple M4, could process blocks of 64 bytes in one go, and with less instructions. But that involves a lot of complications, so I think this is a decent compromise for now.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9359?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9359** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9359?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9359?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9349** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9349?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9348** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9348?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9342** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9342?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9359).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Feb 25, 9:08 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(ast/estree): speed up building UTF8-UTF16 translation table with SIMD (#9359)

Use SIMD operations to speed up building the table of UTF8-UTF16 offset translations.

Previously, we looped through the source byte-by-byte, checking if each byte is ASCII, one-by-one. Instead, use SIMD to check blocks of 32 bytes at a time, and only drop down to byte-by-byte processing if the quick check does find any non-ASCII bytes in that block.

Building the table is about 7x faster in our benchmark after this PR.

Note: There are no explicit SIMD operations involved in the code. Just the code is written so that the compiler can auto-vectorize it.

We could get even better performance if we did use SIMD proper - on x86_64 with AVX-512 or Apple M4, could process blocks of 64 bytes in one go, and with less instructions. But that involves a lot of complications, so I think this is a decent compromise for now.

overlookmotel commented:
Miri passes (see #9370). I think this is good to go.
